### PR TITLE
refactor sample provider code

### PR DESCRIPTION
**Pull request by ngoldman**
Working on refactoring sample provider to be up to date to current Koop standards.

@chelm @dmfenton please feel free to jump in with commits on the `refactor` branch, any help is appreciated and you're both more familiar with the ins and outs of providers

### CODE REVIEW

**Comment by chelm** [comment]
@ngolgman looks pretty good. Nothing really jumps out as needing to updated. We could maybe set a precedent of using jsdocs in the model and controller with an npm script for generating the docs...

**Comment by chelm** [comment]
@ngoldman is it worth adding a `.travis.yml`?

**Comment by chelm** [comment]
We should probably also keep a CHANGELOG.md

**Comment by ngoldman** [comment]
@chelm depends on how we want this to behave. It's a template right now so I think having a CHANGELOG.md with our changes might be weird for someone cloning the project to start a new provider. People also often forget to get rid of that kind of boilerplate when they first commit.

We could write a script that creates a blank template with boilerplate removed, or even have koop generate something... this would be a throwback to that idea we tossed around a while ago about adding some cli functionality to koop.

As far as `.travis.yml` and testing, I think it might make more sense for tests to not pass at first -- basically set up a bunch of failing tests that need to pass for the provider to prove it can work correctly. If we go that route having a .travis.yml for this repo would be superfluous, but would be useful when you're first setting up a new provider.

I definitely like setting the precedent of using jsdoc to document methods. ++

**Comment by chelm** [comment]
@ngoldman I can get behind that. I'll break the tests again...

File: test/model-test.js [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
+    t.ok(data, 'data exists')
+    t.ok(Array.isArray(data), 'data is an array')
+    t.equal(data[0].type, 'FeatureCollection', 'data contains object of type FeatureCollection')
+    t.end()
+  })
+})
 
+test('model: teardown', function (t) {
+  sample.request.restore()
+  t.end()
```
Comment (ngoldman) [inline test/model-test.js]: Noticing the process is hanging when all tests passes. I narrowed it down to the `koop.Cache.insert` call inside of `models/Sample.js`. The callback fires but the process hangs afterwards. Haven't figured out what's causing it yet.

I saw @dmfenton added a `process.exit(0)` at the end of test in `koop-socrata` to deal with this (https://github.com/koopjs/koop-socrata/blob/master/test/model.js#L363) but that only works if that test file is run last (it's possible for the process to exit cleanly after this, skipping a failing test in another file and returning a false positive). Ideally the process shouldn't hang at all. ¯\\\_(ツ)\_/¯

**Comment by ngoldman** [comment]
@jgravois feel free to jump in if you've got ideas for improving conventions in the sample provider or time to rewrite some of the doc in the readme.

File: test/model-test.js (hunk 1 above)
Comment (dmfenton) [inline test/model-test.js]: That won't happen if you explicitly order the specs. like I do in Socrata but its totally a hack.

**Comment by chelm** [comment]
hey @ngoldman When I run `npm test` the tests pass (except still not exiting), but when I run `node test/model-test.js` the test fails. Any idea why?

**Comment by chelm** [comment]
@ngoldman maybe this is by design to have the test fail

**Comment by chelm** [comment]
Seems like the hanging test is associated with the sinon.stub of the sample.request method. Commenting that out and the test exits.

**Comment by chelm** [comment]
ahh nope, cache.insert... i see now

**Comment by chelm** [comment]
This is an issue in the koop modules local cache. I'm going to work on PR and patch.

**Comment by chelm** [comment]
This PR will fix the hanging test issue https://github.com/Esri/koop/pull/203

**Comment by ngoldman** [comment]
cool, thanks for figuring it out @chelm

**Comment by chelm** [comment]
@ngoldman I updated the sample model pattern slightly and the tests are passing. How do you go about making them fail on purpose?

**Comment by ngoldman** [comment]
@chelm I think the idea would be to create tests for the conditions of a working provider and then remove the parts that are faking a working provider

**Comment by ngoldman** [comment]
I think this is good enough for now, we can improve on it as time allows. Definitely better than before :)

Thanks all for the help :+1:

**Pull request merged**